File: main/src/main/java/de/blinkt/openvpn/core/PasswordDialogFragment.java
/*
 * Copyright (c) 2012-2019 Arne Schwabe
 * Distributed under the GNU GPL v2 with additional terms. For full terms see the file doc/LICENSE.txt
 */

package de.blinkt.openvpn.core;

import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.os.RemoteException;
import android.text.InputType;
import android.widget.EditText;


public class PasswordDialogFragment extends DialogFragment {

    private IOpenVPNServiceInternal mService;
    private ServiceConnection mConnection = new ServiceConnection() {


        @Override
        public void onServiceConnected(ComponentName className,
                                       IBinder service) {

            mService = IOpenVPNServiceInternal.Stub.asInterface(service);
        }

        @Override
        public void onServiceDisconnected(ComponentName arg0) {
            mService = null;
        }

    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Intent intent = new Intent(getActivity(), OpenVPNService.class);
        intent.setAction(OpenVPNService.START_SERVICE);
        getActivity().bindService(intent, mConnection, 0);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        getActivity().unbindService(mConnection);
    }

    static PasswordDialogFragment newInstance(String title) {
        PasswordDialogFragment frag = new PasswordDialogFragment();
        Bundle args = new Bundle();
        args.putString("title", title);
        frag.setArguments(args);
        return frag;
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        String title = getArguments().getString("title");

        final EditText input = new EditText(getActivity());
        input.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);


        return new AlertDialog.Builder(getActivity())
                .setIcon(android.R.drawable.ic_dialog_alert)
                .setTitle(title)
                .setView(input)
                .setPositiveButton(android.R.string.ok, (dialogInterface, i) -> {
                    if (mService != null) {
                        try {
                            mService.challengeResponse(input.getText().toString());
                            getActivity().finish();
                        } catch (RemoteException e) {
                            VpnStatus.logException(e);
                            e.printStackTrace();
                        }
                    }
                })
                .setNegativeButton(android.R.string.cancel,
                        (dialog, whichButton) -> getActivity().finish()
                )
                .create();
    }
}

File: main/src/ui/java/de/blinkt/openvpn/fragments/PasswordDialogFragment.kt
/*
 * Copyright (c) 2012-2019 Arne Schwabe
 * Distributed under the GNU GPL v2 with additional terms. For full terms see the file doc/LICENSE.txt
 */
package de.blinkt.openvpn.core

import android.app.AlertDialog
import android.app.Dialog
import android.content.ComponentName
import android.content.DialogInterface
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.os.RemoteException
import android.text.InputType
import android.widget.EditText
import androidx.fragment.app.DialogFragment
import de.blinkt.openvpn.R
import de.blinkt.openvpn.core.OpenVPNService.EXTRA_CHALLENGE_TXT

class PasswordDialogFragment : DialogFragment() {
    private var mService: IOpenVPNServiceInternal? = null
    private val mConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName,
                                        service: IBinder) {
            mService = IOpenVPNServiceInternal.Stub.asInterface(service)
        }

        override fun onServiceDisconnected(arg0: ComponentName) {
            mService = null
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val intent = Intent(activity, OpenVPNService::class.java)
        intent.action = OpenVPNService.START_SERVICE
        requireActivity().bindService(intent, mConnection, 0)
    }

    override fun onDestroy() {
        super.onDestroy()
        requireActivity().unbindService(mConnection)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val title = requireArguments().getString("title")
        val echo = requireArguments().getBoolean("echo")
        val finish = requireArguments().getBoolean("finish")
        val input = EditText(activity)
        if (!echo)
            input.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD

        return AlertDialog.Builder(activity)
                .setTitle("Challenge/Response Authentification")
                .setMessage(title)
                .setView(input)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    try {
                        mService?.challengeResponse(input.text.toString())
                        if (finish) requireActivity().finish()
                    } catch (e: RemoteException) {
                        VpnStatus.logException(e)
                        e.printStackTrace()
                    }
                }
                .setNegativeButton(R.string.cancel
                ) { _, _ -> if (finish) requireActivity().finish() }
                .create()
    }

    companion object {
        fun newInstance(intent: Intent, finish: Boolean): PasswordDialogFragment? {
            val extras = intent.extras ?: return null
            val challenge = extras.getString(EXTRA_CHALLENGE_TXT, "R,E:(empty challenge text)")
            val message = challenge.split(":", limit = 2)[1]
            val flagsStr = challenge.split(":", limit = 2)[0]
            val flags = flagsStr.split(",")
            var echo = false
            var response = false
            for (flag in flags) {
                if (flag == "R") response = true
                else if (flag == "E") echo = true
            }
            if (!response) {
                VpnStatus.logError("Error unrecognised challenge from Server: $challenge")
                return null
            }
            val frag = PasswordDialogFragment()
            val args = Bundle()
            args.putString("title", message)
            args.putBoolean("echo", echo)
            args.putBoolean("finish", finish)
            frag.arguments = args
            return frag
        }
    }
}